### PR TITLE
events: add account_id while creating client in sending events to target

### PR DESCRIPTION
**Pull request by sannya-singal**
## Motivation
When using values other than `000000000000` for account id or `us-east-1` for region, `events` tests should still create the consequent resources in the corresponding accounts and region.

## Changes
This PR introduces the use of `account_id` field fetched from `target_arn` in order to connect with the correct client in `connect_to()` method while sending events to the target.

## Testing

The tests were failing previously when executed with a non-default account ID, set through environment variables (`TEST_AWS_ACCOUNT_ID=111111111111`, `TEST_AWS_ACCESS_KEY_ID=111111111111`, `TEST_AWS_REGION=us-west-1`). The affected tests are:

```
tests.aws.services.events.test_events.TestEvents.test_put_events_into_event_bus[standard]
tests.aws.services.events.test_events.TestEvents.test_put_events_into_event_bus[domain]
tests.aws.services.events.test_events.TestEvents.test_put_events_into_event_bus[path]
tests.aws.services.events.test_events.TestEvents.test_put_events_with_target_firehose
tests.aws.services.events.test_events.TestEvents.test_put_events_with_target_kinesis
tests.aws.services.cloudformation.resources.test_events.test_event_rule_to_logs
```

### CODE REVIEW

**Review by viren-nadkarni — DISMISSED** [review]
Similar to the previous PRs, the proper fix here is to make sure IAM setup is performed in the tests.

In this case, you must create and pass an appropriate role in the `PutRule` or `PutTarget` calls in these tests.

https://github.com/localstack/localstack/blob/f4e9632f4ff2bff6514f8fd42c631e0db510e204/tests/aws/services/events/test_events.py#L530-L534

Refer to https://docs.aws.amazon.com/eventbridge/latest/APIReference/API_PutTargets.html

**Comment by sannya-singal** [comment]
> this case, you must create and pass an appropriate role in the `PutRule` call in these tests.

In case of there is no IAM enforcement the setup should still work, right? 🤔 @viren-nadkarni?

**Comment by sannya-singal** [comment]
According to my understanding in case of the [previous PR](https://github.com/localstack/localstack/pull/9639#discussion_r1394128323) we were explicitly setting the `access_key_id` as `None` but in this case we are already using `connect_to()` with `region` which will lead it to connect with the correct region and incorrect `account_id` even when the role is not specified.

**Comment by viren-nadkarni** [comment]
@sannya-singal Please also make sure that the tests you're trying to address pass in the CI with a commit like: b21b9d86a1fc6a886fb04ac8f8f9c3f3807ad561 which you can revert before merge

**Comment by sannya-singal** [comment]
> The test is not AWS validated, so we can't make that assumption. Even if IAM enforcement is disabled, role assumption is still a proper way to access resources.

@viren-nadkarni 

Maybe lets wait for review from @dfangl because its a bit confusing due to a comment [here](https://github.com/localstack/localstack/pull/9639#pullrequestreview-1744449202) which says it has to work without a role also, additionally we can add some more test cases for IAM enforcement to increase the robustness of multi-account and region functionality. 

Will also run a CI pipeline for non-default `account_id` and `region`.

**Review by dfangl — APPROVED** [review]
Like discussed, we need to probably look at it in a follow up!

**Comment by sannya-singal** [comment]
Failing ext integration tests pipeline- https://github.com/localstack/localstack-ext/actions/runs/7738980761/job/21144883663

Failing tests: 

```
FAILED tests/aws/services/iam/test_inter_service_enforcement.py::TestInterServiceCommunicationEnforcement::test_events_sqs - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED tests/aws/services/iam/test_inter_service_enforcement.py::TestInterServiceCommunicationEnforcement::test_events_lambda - Failed: DID NOT RAISE <class 'botocore.exceptions.ClientError'>
FAILED tests/aws/services/iam/test_inter_service_enforcement.py::TestInterServiceCommunicationEnforcement::test_events_sns - Failed: DID NOT RAISE <class 'AssertionError'>
FAILED tests/aws/services/iam/test_inter_service_enforcement.py::TestInterServiceCommunicationEnforcement::test_events_logs - Failed: DID NOT RAISE <class 'AssertionError'>
```

File: localstack/utils/aws/message_forwarding.py
Comment (viren-nadkarni) [inline localstack/utils/aws/message_forwarding.py:60]: Are you sure about introducing this circular dependency? For happy path, shouldn't it be ideal to pass the role?

Comment (sannya-singal) [reply localstack/utils/aws/message_forwarding.py:60]: Passing the role param here was already discussed and approved earlier. The new fix of introducing the `ENFORCE_IAM` attribute accounts to handle the failing ext test cases as mentioned [here](https://github.com/localstack/localstack/pull/9784#issuecomment-1938155138). Waiting for @dfangl review on this.

Comment (dfangl) [inline localstack/utils/aws/message_forwarding.py:60]: I am not a fan of this at all. We should never make client creation dependent on the enforce IAM flag set - it should always be configured identically, with error handling if it fails due to a missing IAM permission. We should also not import pro code from community.

We have to ask us why this is necessary here. For all services which allow resource based policies (like lambda, sqs) it should not be necessary - they should parse the arn from the target themselves.

For those where it is necessary, it should always be conceptually the source account, not the target one (if cross account is possible).
We can allow this in the IAM engine itself - I can open a PR for it. Currently, it will recognize a call as being made from a service principal only if the access key id is `__internal__` - if this is too limiting, we can accept numeric account ids there as well.

File: localstack/utils/aws/message_forwarding.py
Comment (dfangl) [inline localstack/utils/aws/message_forwarding.py:42]: We should use the source arn here conceptually - not sure if cross account is allowed here, but if it is, the source arn would be the way to go. Services that do not allow cross account requests (like firehose) would not work, but that would be fine.

**Review by dfangl — CHANGES_REQUESTED** [review]
I am not a fan of this ext import here at all - the client creation should always be identical, and always contain all information necessary for IAM enforcement.

I detailed it in the comments, and will open up a PR in -ext today.

**Review by dfangl — APPROVED** [review]
If all the tests pass, the changes look good :)

**Comment by sannya-singal** [comment]
Ext Integration Tests: https://github.com/localstack/localstack-ext/actions/runs/7927487686

**Pull request merged**
